example/mailto/mailto.cpp

```diff
@@ -0,0 +1,377 @@
+
+#include "mailto_grammar.hpp"
+#include <boost/url/url_view.hpp>
+#include <boost/url/url.hpp>
+#include <boost/url/optional.hpp>
+#include <boost/url/grammar/ci_string.hpp>
+#include <boost/url/parse.hpp>
+#include <boost/url/rfc/absolute_uri_rule.hpp>
+#include <boost/url/grammar/digit_chars.hpp>
+#include <boost/url/grammar/parse.hpp>
+#include <boost/url/grammar/vchars.hpp>
+#include <boost/url/grammar/alpha_chars.hpp>
+#include <boost/url/grammar/alnum_chars.hpp>
+#include <boost/url/grammar/token_rule.hpp>
+#include <boost/url/grammar/alnum_chars.hpp>
+#include <iostream>
+#include <algorithm>
+
+namespace urls = boost::urls;
+
+// fwd-declaration for mailto_view
+struct mailto_rule_t;
+
+/// A new url type for mailto URLs
+/**
+    This class represents a URI with the mailto
+    scheme.
+
+    Unlike a urls::url_view, which only represents
+    the general syntax of urls, a mailto_view
+    represents a reference to fields that are
+    relevant to mailto URLs, while ignoring
+    elements of the general syntax
+    that are not relevant to the scheme.
+
+    This allows us to use the general syntax
+    parsers to create a representation that
+    is more appropriate for the specified scheme
+    syntax.
+
+    @par Specification
+    @li <a href="https://www.rfc-editor.org/rfc/rfc6068"
+        >The 'mailto' URI Scheme</a>
+    @li <a href="https://www.rfc-editor.org/rfc/rfc2368"
+        >The mailto URL scheme</a>
+    @li <a href="https://www.rfc-editor.org/rfc/rfc1738"
+        >Uniform Resource Locators (URL)</a>
+
+    @par References
+    @li <a href="https://en.wikipedia.org/wiki/Mailto"
+        >mailto (Wikipedia)</a>
+
+ */
+class mailto_view
+{
+    urls::url_view u_;
+
+public:
+    /// Return the specified email address in the URL
+    /**
+        A mailto URL might contain multiple email
+        addresses separated by commas.
+
+        The first addresses are represented in
+        the path. Other addresses are in
+        any query parameter whose key is "to".
+
+        @param idx Address index
+
+        @return The specified address
+     */
+    std::string
+    address(std::size_t idx = 0) const
+    {
+        return encoded_address(idx).decode();
+    }
+
+    /// @copydoc address()
+    urls::pct_string_view
+    encoded_address(std::size_t idx = 0) const noexcept
+    {
+        // Look for ith email address in the path string
+        auto s = ith_address(u_.encoded_path(), idx);
+        if (s)
+            return *s;
+
+        // Look for ith email address in one of the "to" headers
+        auto ps = u_.encoded_params();
+        auto it = ps.find("to", urls::ignore_case);
+        while (it != ps.end())
+        {
+            s = ith_address((*it++).value, idx);
+            if (s)
+                return *s;
+            it = ps.find(it, "to", urls::ignore_case);
+        }
+        return {};
+    }
+
+    /// Return number of email addresses in the URL
+    std::size_t
+    size() const noexcept
```

Comment:
We need to separate all the function definitions from the declarations, because otherwise there is no good way to look at the synopsis for the class to get a feel for its API. Sorry!